### ISSUE TITLE
Pyright does not evaluate type correctly when calling a generic class constructor with a union of classes

### ISSUE THREAD

**Issue by math-fehr** [microsoft/pyright#5022]
**Describe the bug**
When subclass of `Generic[T]` is expecting a `type[T]` in its `__init__`,
passing a union type prevents the inference of `T`.

**To Reproduce**

With pyright 1.1.304:

```python
from __future__ import annotations
from dataclasses import dataclass
from typing import Generic, TypeVar
from typing_extensions import reveal_type


T = TypeVar("T")


@dataclass
class Foo(Generic[T]):
    t: type[T]


reveal_type(Foo(int))  # Type of "Foo(int)" is "Foo[int]"
reveal_type(Foo(int | bool))  # Type of "Foo(int | bool)" is "Foo[Unknown]"
```

This behavior is different with functions, which has the behavior I
expected.

```python
from __future__ import annotations
from dataclasses import dataclass
from typing import Generic, TypeVar
from typing_extensions import reveal_type

T = TypeVar("T")

def foo(t: type[T]) -> Foo[T]:
    raise NotImplementedError()


reveal_type(foo(int))  # Type of "foo(int)" is "Foo[int]"
reveal_type(foo(int | bool))  # Type of "foo(int | bool)" is "Foo[int | bool]"
```

**Expected behavior**
I would expect that `Foo(int | bool)` would be inferred
to be a `Foo[int | bool]`, but instead it got inferred to `Foo[Unknown]`.

**VS Code extension or command-line**
I get this behavior in pyright 1.1.304 both in the VS Code extension
and in the command line.

**Additional context**
I also tried to see what MyPy outputs, and it gives me an error
for trying to use an `UnionType` in a `type[...]`, both in the class
example and the function example. I'm not sure if this is a limitation
of MyPy, or an error in pyright.
```
Argument 1 to "Foo" has incompatible type "UnionType"; expected "Type[<nothing>]"
Argument 1 to "foo" has incompatible type "UnionType"; expected "Type[<nothing>]"
```

**Comment by erictraut** [comment]
You're using a union operator (`|`) in a value expression — i.e. an expression that is evaluated at runtime and passed as an argument. Typically, the union operator is used only in type annotations and is not evaluated at runtime. When it's evaluated at runtime, it produces an object that is an instance of the `UnionType` class.

Perhaps your intent was to do the following?

```python
def func(val: int | bool):
    reveal_type(Foo(type(val)))
```

Or this?

```python
@dataclass
class Foo(Generic[T]):
    t: type[T] | tuple[type[T], ...]

reveal_type(Foo((int, bool)))
```

**Comment by math-fehr** [comment]
Oh, thanks a lot, the second example seems to do exactly what I needed!

About the function example I gave, should I understand that pyright is wrong by allowing `foo(int | bool)`, as this is not of type `type[T]`, but of type `UnionType`?

**Comment by erictraut** [comment]
Yes, pyright should emit errors in these cases, as mypy does. I'll add that check.

**Comment by erictraut** [comment]
@mhc-cs-mdrissi, that's a good point. Pyright differs from mypy in this regard (although mypy is arguably following the PEP 484 spec more faithfully). But I agree that until we have a `TypeForm` or something similar, adding an error here would eliminate an interim solution for design patterns like the one you mentioned.

I've already written the code (and tests) for generating errors in this case, but I'll refrain from checking it in. 

Instead, I'll fix the inconsistency that @math-fehr originally reported with the constructor call. If call to the generic function works, then the call to the generic class constructor should be consistent.

**Comment by math-fehr** [comment]
Having now read the `TypeForm` issue, it seems that what I needed is exactly `TypeForm` (and I also have used `type[T]` when I actually needed `TypeForm[T]` in other parts of code).
Thanks a lot for keeping that feature until `TypeForm` is ready!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.306, which I just published. It will also be included in a future release of pylance.
